forester/src/batched_ops.rs

```diff
@@ -0,0 +1,481 @@
+
+            let response = client
+                .post(&format!("{}{}", SERVER_ADDRESS, PROVE_PATH))
+                .header("Content-Type", "text/plain; charset=utf-8")
+                .body(inputs_json)
+                .send()
+                .await
+                .expect("Failed to execute request.");
+
+            if response.status().is_success() {
+                let body = response.text().await.unwrap();
+                let proof_json = deserialize_gnark_proof_json(&body).unwrap();
+                let (proof_a, proof_b, proof_c) = proof_from_json_struct(proof_json);
+                let (proof_a, proof_b, proof_c) = compress_proof(&proof_a, &proof_b, &proof_c);
+                (
+                    CompressedProof {
+                        a: proof_a,
+                        b: proof_b,
+                        c: proof_c,
+                    },
+                    bigint_to_be_bytes_array::<32>(&circuit_inputs.new_root.to_biguint().unwrap())
+                        .unwrap(),
+                )
+            } else {
+                error!(
+                    "create_append_batch_ix_data: failed to get proof from server: {:?}",
+                    response.text().await
+                );
+                return Err(ForesterError::Custom(
+                    "Failed to get proof from server".into(),
+                ));
+            }
+        };
+
+        info!("=== create_append_batch_ix_data end ===");
+
+        Ok(InstructionDataBatchAppendInputs {
+            public_inputs: AppendBatchProofInputsIx { new_root },
+            compressed_proof: proof,
+        })
+    }
+
+    async fn get_batched_nullify_ix_data(&self) -> Result<InstructionDataBatchNullifyInputs> {
+        info!("=== get_batched_nullify_ix_data begin ===");
+        let mut rpc = self.rpc_pool.get_connection().await.unwrap();
+
+        let (zkp_batch_size, batch_index, old_root, old_root_index, leaves_hashchain) = {
+            let mut account = rpc.get_account(self.merkle_tree).await.unwrap().unwrap();
+            let merkle_tree =
+                ZeroCopyBatchedMerkleTreeAccount::from_bytes_mut(account.data.as_mut_slice())
+                    .unwrap();
+            let account = merkle_tree.get_account();
+            let batch_idx = account.queue.next_full_batch_index as usize;
+            let zkp_size = account.queue.zkp_batch_size;
+            let batch = &merkle_tree.batches[batch_idx];
+            let zkp_idx = batch.get_num_inserted_zkps();
+            let hashchain = merkle_tree.hashchain_store[batch_idx][zkp_idx as usize];
+            let root_idx = merkle_tree.root_history.last_index();
+            let root = *merkle_tree.root_history.last().unwrap();
+
+            (zkp_size, batch_idx, root, root_idx, hashchain)
+        };
+
+        let leaf_indices_tx_hashes = self
+            .indexer
+            .lock()
+            .await
+            .get_leaf_indices_tx_hashes(self.merkle_tree, zkp_batch_size as usize);
```

Comment:
wondering whether it would make sense to return `get_proof_by_index(self.merkle_tree, *index as usize);` with this call as well so that we avoid the second call to the indexer.